src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/ConstantExpectedTests.cs

```diff
@@ -0,0 +1,722 @@
+        [InlineData("uint", "3", "4", "5", 3, 4)]
+        [InlineData("ulong", "3", "4", "5", 3, 4)]
+        [InlineData("nuint", "3", "4", "5", 3, 4)]
+        [InlineData("float", "3", "4", "5", 3, 4)]
+        [InlineData("double", "3", "4", "5", 3, 4)]
+        [InlineData("bool", "false", "false", "true", false, false)]
+        public static async Task TestConstantCompositionOutOfBoundsAsync(string type, string min, string max, string outOfBoundMax, object minValue, object maxValue)
+        {
+            string csInput = @$"
+using System;
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public class Test
+{{
+    public static void TestMethodConstrained([ConstantExpected(Min = {min}, Max = {outOfBoundMax})] {type} constant)
+    {{
+        TestMethodWithConstrainedConstant({{|#0:constant|}});
+    }}
+    public static void TestMethodWithConstrainedConstant([ConstantExpected(Min = {min}, Max = {max})] {type} val) {{ }}
+}}
+";
+            await TestCSAsync(csInput,
+                VerifyCS.Diagnostic(ConstantExpectedAnalyzer.CA1861.ConstantOutOfBoundsRule)
+                        .WithLocation(0)
+                        .WithArguments(minValue.ToString(), maxValue.ToString()));
+        }
+
+        [Fact]
+        public static async Task TestConstantCompositionNotSameTypeAsync()
+        {
+            string csInput = @"
+using System;
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public class Test
+{
+    public static void TestMethod([ConstantExpected] long constant)
+    {
+        TestMethodWithConstant({|#0:(int)constant|});
+        TestMethodWithStringConstant({|#1:(string)(object)constant|});
+    }
+    public static void TestMethod([ConstantExpected] short constant)
+    {
+        TestMethodWithConstant({|#2:constant|});
+    }
+    public static void TestMethodWithConstant([ConstantExpected] int val) { }
+    public static void TestMethodWithStringConstant([ConstantExpected] string val) { }
+}
+";
+            await TestCSAsync(csInput,
+                VerifyCS.Diagnostic(ConstantExpectedAnalyzer.CA1861.ConstantInvalidConstantRule)
+                        .WithLocation(0)
+                        .WithArguments("int"),
+                VerifyCS.Diagnostic(ConstantExpectedAnalyzer.CA1861.ConstantInvalidConstantRule)
+                    .WithLocation(1)
+                    .WithArguments("string"),
+                VerifyCS.Diagnostic(ConstantExpectedAnalyzer.CA1861.ConstantInvalidConstantRule)
+                        .WithLocation(2)
+                        .WithArguments("int"));
+        }
+
+        private static async Task TestCSAsync(string source, params DiagnosticResult[] diagnosticResults)
+        {
+            var test = new VerifyCS.Test
+            {
+                TestCode = source,
+                LanguageVersion = CSharp.LanguageVersion.Preview,
+            };
+            // TODO: remove when the type is available
```

Comment:
```suggestion
            // TODO: remove when the type is available
```
which type?

Reply:
the attribute source being added. There isn't a Net70 reference assembly for the test yet